File: backend/Model/TaskManagement/StaffTask.cs
namespace YLWorks.Model
{
    public class StaffTask : BaseEntity
    {
        public Guid Id { get; set; }

        public string Title { get; set; } = null!;

        public string? Description { get; set; }

        public Guid? AssignedToId { get; set; }

        public User? AssignedTo { get; set; } = null!;

        public Guid AssignedById { get; set; }

        public User AssignedBy { get; set; } = null!;

        public string Priority { get; set; } = "Medium";

        // Admin, HR, Development, Documentation,
        // Maintenance, Meeting, Training, Others
        public string Category { get; set; } = "Others";

        // Todo, InProgress, Pending, Completed, Cancelled
        public string Status { get; set; } = "Todo";

        public DateTime? StartDate { get; set; }

        public DateTime DueDate { get; set; }

        public DateTime? ReminderAt { get; set; }

        public bool IsRecurring { get; set; }

        // Daily, Weekly, Monthly, Yearly
        public string? RecurringType { get; set; }

        public decimal? EstimatedHours { get; set; }

        public decimal? ActualHours { get; set; }

        public DateTime? CompletedAt { get; set; }

        public Guid? CompletedById { get; set; }

        public User? CompletedBy { get; set; }

        public ICollection<StaffTaskChecklist> Checklists { get; set; }
            = new List<StaffTaskChecklist>();
    }


    public class StaffTaskChecklist
    {
        public Guid Id { get; set; }

        public Guid StaffTaskId { get; set; }

        public StaffTask StaffTask { get; set; } = null!;

        public string Title { get; set; } = null!;

        public bool IsCompleted { get; set; }

        public DateTime? CompletedAt { get; set; }

        public int Sequence { get; set; }
    }

    public class CreateStaffTaskRequest
    {
        public string Title { get; set; } = null!;

        public string? Description { get; set; }

        public Guid? AssignedToId { get; set; }

        public string Priority { get; set; } = "Medium";

        public string Category { get; set; } = "Others";

        public DateTime? StartDate { get; set; }

        public DateTime DueDate { get; set; }

        public DateTime? ReminderAt { get; set; }

        public bool IsRecurring { get; set; }

        public string? RecurringType { get; set; }

        public decimal? EstimatedHours { get; set; }

        public List<CreateStaffTaskChecklistRequest> Checklists { get; set; }
            = new();
    }


    public class CreateStaffTaskChecklistRequest
    {
        public string Title { get; set; } = null!;

        public int Sequence { get; set; }
    }


    public class UpdateStaffTaskRequest
    {
        public Guid Id { get; set; }

        public string Title { get; set; } = null!;

        public string? Description { get; set; }

        public Guid? AssignedToId { get; set; }

        public string Priority { get; set; } = "Medium";

        public string Category { get; set; } = "Others";

        public string Status { get; set; } = "Todo";

        public DateTime? StartDate { get; set; }

        public DateTime DueDate { get; set; }

        public DateTime? ReminderAt { get; set; }

        public bool IsRecurring { get; set; }

        public string? RecurringType { get; set; }

        public decimal? EstimatedHours { get; set; }

        public List<UpdateStaffTaskChecklistItemRequest> Checklists { get; set; }
            = new();
    }


    // Used when creator updates the whole task
    public class UpdateStaffTaskChecklistItemRequest
    {
        public Guid? Id { get; set; }

        public string Title { get; set; } = null!;

        public bool IsCompleted { get; set; }

        public int Sequence { get; set; }
    }


    // Used ONLY by assigned staff when ticking checklist
    public class UpdateStaffTaskChecklistRequest
    {
        public Guid Id { get; set; }

        public List<UpdateStaffTaskChecklistItemOnlyRequest> Checklists { get; set; }
            = new();
    }


    public class UpdateStaffTaskChecklistItemOnlyRequest
    {
        public Guid Id { get; set; }

        public bool IsCompleted { get; set; }

        public int Sequence { get; set; }
    }


    public class CompleteStaffTaskRequest
    {
        public decimal? ActualHours { get; set; }
    }
}

File: frontend/src/app/models/StaffTask.ts
import { BaseModel } from './BaseModel';
import { UserDto } from './User';

export interface StaffTask extends BaseModel {
  title: string;
  description?: string;
  assignedToId: string;
  assignedTo: UserDto;
  assignedById: string;
  assignedBy: UserDto;
  priority: string;
  category: string;
  status: string;
  startDate: Date;
  dueDate: Date;
  reminderAt?: Date;
  isRecurring: boolean;
  recurringType?: string;
  estimatedHours?: number;
  actualHours?: number;
  completedAt?: Date;
  completedById?: string;
  completedBy?: UserDto;
  checklists: StaffTaskChecklist[];
}

export interface StaffTaskChecklist {
  id: string;
  staffTaskId: string;
  staffTask: StaffTask;
  title: string;
  isCompleted: boolean;
  completedAt?: Date;
  sequence?: number;
}

export interface CreateStaffTaskRequest {
  title: string;
  description?: string;
  assignedToId?: string;
  priority: string;
  category: string;
  startDate?: Date;
  dueDate: Date;
  reminderAt?: Date;
  isRecurring: boolean;
  recurringType?: string;
  estimatedHours?: number;
  checklists: CreateStaffTaskChecklistRequest[];
}

export interface CreateStaffTaskChecklistRequest {
  title: string;
  sequence: number;
}

export interface UpdateStaffTaskRequest {
  id: string;
  title: string;
  description?: string;
  assignedToId: string;
  priority: string;
  category: string;
  status: string;
  startDate?: Date;
  dueDate: Date;
  reminderAt?: Date;
  isRecurring: boolean;
  recurringType?: string;
  estimatedHours?: number;
  checklists: UpdateStaffTaskChecklistItem[];
}

export interface UpdateStaffTaskChecklistItem {
  id: string;
  title: string;
  isCompleted: boolean;
  sequence: number;
}

export interface UpdateChecklistOnlyRequest {
  id: string;
  checklists: UpdateChecklistOnlyItem[];
}

export interface UpdateChecklistOnlyItem {
  id: string;
  isCompleted: boolean;
  sequence: number;
}
export interface UpdateChecklistItemRequest {
  id: string;
  isCompleted: boolean;
  sequence: number;
}
export interface CompleteStaffTaskRequest {
  actualHours?: number;
}

export interface StaffTaskSummary {
  myTasks: number;
  assignedToOthers: number;
  inProgress: number;
  completed: number;
}
